### PR TITLE
First draft to generate tv via run/tv/run script

### PR DESCRIPTION
**Pull request by abhilashendurthi**

### CODE REVIEW

**Comment by abhilashendurthi** [comment]
Draft consists of:
1. run/tv/run and supporting scripts to generate testvectors without running ptf tests. 
   Usage: run/tv/run <profile> [test-case] [device] [portmap] [grpc-addr]
2. run/tv/setup_interfaces.sh to create interfaces in the test runner container based on portmap file.
3. Changes to ptf_runner.py, base_test.py supporting if/else condition to generate either test vectors or run ptf tests

File: tests/ptf/ptf_runner.py [hunk 1]
```diff
@@ -197,21 +200,33 @@ def run_test(p4info_path, grpc_addr, device_id, cpu_port, ptfdir, port_map_path,
...
-
-    # Write the portmap proto object to testvectors/portmap.pb.txt
-    pmutils.write_to_file(tv_portmap, os.getcwd())
+            if generate_tv:
+                # Append new entry to tv proto object
+                interfaces = interfaces + " " + iface_name
+                pmutils.add_new_entry(tv_portmap, p4_port, iface_name)
+    if generate_tv:
+        try:
+            cmd = os.getcwd() + "/../../run/tv/setup_interfaces.sh" + interfaces
```
Comment (ccascone) [inline tests/ptf/ptf_runner.py:219]: Please add a comment to explain why you're setting up interfaces

File: run/tv/Makefile [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+$(if $(filter $(DEVICE),tofino),
+python -u ptf_runner.py --device $(DEVICE) --port-map $(PORTMAP) \
+	--ptf-dir fabric.ptf --cpu-port 320 --device-id 1 \
+	--grpc-addr $(GRPCADDR) \
+	--p4info $(tofino_p4c_out)/$(1)/tofino/mavericks_sde_9_0_0/p4info.txt \
+	--tofino-bin $(tofino_p4c_out)/$(1)/tofino/mavericks_sde_9_0_0/pipe/tofino.bin \
+	--tofino-ctx-json $(tofino_p4c_out)/$(1)//tofino/mavericks_sde_9_0_0/pipe/context.json \
+	--generate-tv \
+	$(2),
+python -u ptf_runner.py --device $(DEVICE) --port-map $(PORTMAP) \
```
[the remark this answers is not in the document]
Comment (ccascone) [reply run/tv/Makefile:14]: Never mind, I didn't see the $(if $(filter $(DEVICE),tofino),

File: run/tv/Makefile [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	--ptf-dir fabric.ptf --cpu-port 255 --device-id 1 \
+	--grpc-addr 127.0.0.1:28000 \
```
Comment (ccascone) [inline run/tv/Makefile:16]:
```suggestion
	--grpc-addr $(GRPCADDR) \
```

File: run/tv/Makefile [hunk 4]
```diff
@@ -0,0 +1,43 @@
+bmv2_p4c_out := ${ONOS_ROOT}/pipelines/fabric/impl/src/main/resources/p4c-out
+tofino_p4c_out := ${FABRIC_TOFINO}/src/main/resources/p4c-out
+
+define run_tests
+$(if $(filter $(DEVICE),tofino),
+python -u ptf_runner.py --device $(DEVICE) --port-map $(PORTMAP) \
+	--ptf-dir fabric.ptf --cpu-port 320 --device-id 1 \
```
Comment (ccascone) [inline run/tv/Makefile:7]: The CPU port should be parametric

File: run/tv/Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (abhilashendurthi) [reply run/tv/Makefile:14]: In order to generate testvectors for both bmv2 and tofino targets, we need two ptf_runner.py calls. Based on the user input on the type of device, we invoke one of the ptf_runner.py calls. I will add a comment explaining the same.

File: run/tv/Makefile [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	--grpc-addr $(GRPCADDR) \
+	--p4info $(tofino_p4c_out)/$(1)/tofino/mavericks_sde_9_0_0/p4info.txt \
```
Comment (ccascone) [inline run/tv/Makefile:9]: This implementation will only work for mavericks and compiler output for SDE 9.0.0. Is there a way to make this parametric? I think simply allowing to pass the full location of the compiler output should work. I.e., setting `$(tofino_p4c_out)` to `/$(1)/tofino/mavericks_sde_9_0_0/`

File: tests/ptf/ptf_runner.py [hunk 6]
```diff
@@ -87,108 +87,111 @@ def build_tofino_config(prog_name, bin_path, cxt_json_path):
...
-            stub.SetForwardingPipelineConfig(request)
-        except Exception as e:
-            error("Error during SetForwardingPipelineConfig")
-            error(str(e))
-            return False
         return True
-    finally:
-        stream_out_q.put(None)
-        stream_recv_thread.join()
+    else:
```
[the remark this answers is not in the document]
Comment (abhilashendurthi) [reply tests/ptf/ptf_runner.py]: My bad, sorry for not doing that earlier. Fixed now.

File: tests/ptf/ptf_runner.py (hunk 1 above)
Comment (abhilashendurthi) [reply tests/ptf/ptf_runner.py:223]: Done.

File: README.md [hunk 7]
```diff
@@ -184,9 +184,82 @@ to generate protobuf-based [TestVectors] (TVs) (under `tests/ptf/testvectors`).
...
+   Example command with all the optional arguments set:
+
+   ```
+   ./run/tv/run fabric DEVICE=tofino PORTMAP=port_map.hw.json GRPCADDR=10.128.13.111:28000 CPUPORT=320 TEST=test.FabricBridgingTest
+   ```
+
+    **NOTE:** Testing `all` profiles is not supported on this target. You must
+    execute the run command for each profile.
+
+    To run all test cases for the basic `fabric` profile:
```
Comment (ccascone) [inline README.md:250]:
```suggestion
    To generate TestVectors for the basic `fabric` profile:
```

File: README.md [hunk 8]
```diff
@@ -184,9 +184,82 @@ to generate protobuf-based [TestVectors] (TVs) (under `tests/ptf/testvectors`).
...
+
+    ```
+    ./run/tv/run fabric
+    ```
+
+    To run a specific test case against a specific fabric profile,
```
Comment (ccascone) [inline README.md:256]:
```suggestion
    To generate a specific test case for a specific fabric profile,
```

**Pull request merged**